# Lab 9
## Submitted By : Aagnay Kariyal
## Student ID : 8830232

***

### Data loading and exploration

In [48]:
# Importing all the necessary dependencies
import numpy as np
from keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from sklearn.metrics import confusion_matrix

In [39]:
# Loading the data and splitting them into variables
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

[Click here for the official Keras documentation of fMNIST](https://keras.io/keras_core/api/datasets/fashion_mnist/)

From the main keras website, we can see that X_Train and X_Test data consists of image data and y_train and y_test data consists of their respective labels.

In [75]:
# Data Exploration
print('Shapes of data: \n')
print(f'X Train: {X_train.shape}')
print(f'X Test: {X_test.shape}')
print(f'Y Train: {y_train.shape}')
print(f'Y Test: {y_test.shape}')

Shapes of data: 

X Train: (60000, 28, 28)
X Test: (10000, 28, 28)
Y Train: (60000,)
Y Test: (10000,)


After a brief exploration of the data, we were able to identify that the Train data consists of 60,000 image data and their respective labels and the Train data consists of 10,000 image data and their respective labels as well.

From the documentation, we were also able to identify that the labels are represented in the form of integers naming from 0-9

### Training a dense neural network using Keras

***


In [42]:
# Preprocessing the data
train_data = X_train.reshape((60000, 28 * 28))
test_data = X_test.reshape((10000, 28 * 28))

train_data = train_data.astype('float32') / 255
test_data = test_data.astype('float32') / 255

In [44]:
parameter = [128, 130, 240] # Setting up some hyperparameters to iterate through to get the best configuration
accuracy = {} # Defining a dictionary to save the accuracy and hyperparameters of each model

# Creating a for loop to iterate through each model 
for p in parameter:
    model = Sequential([
        Dense(p, activation='relu'),
        Dense(10, activation='softmax')
    ])
    
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    
    model.fit(train_data, y_train, epochs=10, batch_size=128)
    
    loss_of_data, accuracy_of_data = model.evaluate(test_data, y_test)
    
    accuracy[p] = accuracy_of_data  # Storing all the accuracy values and the hyperparameter values into the dictionary

predict = model.predict(test_data)
pred_labels = np.argmax(predict, axis=1)

Epoch 1/10
469/469 [==============================] - 14s 21ms/step - loss: 0.5870 - accuracy: 0.7986
Epoch 2/10
469/469 [==============================] - 9s 18ms/step - loss: 0.4563 - accuracy: 0.8410
Epoch 3/10
469/469 [==============================] - 6s 12ms/step - loss: 0.4371 - accuracy: 0.8496
Epoch 4/10
469/469 [==============================] - 10s 21ms/step - loss: 0.4280 - accuracy: 0.8508
Epoch 5/10
469/469 [==============================] - 5s 11ms/step - loss: 0.4244 - accuracy: 0.8524
Epoch 6/10
469/469 [==============================] - 5s 11ms/step - loss: 0.4196 - accuracy: 0.8544
Epoch 7/10
469/469 [==============================] - 6s 14ms/step - loss: 0.4189 - accuracy: 0.8544
Epoch 8/10
469/469 [==============================] - 7s 15ms/step - loss: 0.4175 - accuracy: 0.8545
Epoch 9/10
469/469 [==============================] - 5s 11ms/step - loss: 0.4181 - accuracy: 0.8549
Epoch 10/10
313/313 [==============================] - 7s 13ms/step - loss: 0.4601 - accu

### Accuracy Evaluation, Confusion metric and Incorrect Classification examples

***


In [77]:
print(accuracy)

{128: 0.8403000235557556, 130: 0.8410999774932861, 240: 0.8338000178337097}


From the above data, we can see that the model configuration with the hyperparameter 130 gave the highest accuracy of all three models

In [76]:
cm = confusion_matrix(y_test, pred_labels)
print(cm)

[[726  12  12  97   6   2 137   0   7   1]
 [  1 961   0  32   3   0   2   0   1   0]
 [ 12  11 601  15 184   1 171   0   5   0]
 [  7   8   9 908  26   1  38   0   3   0]
 [  0   8  45  49 779   1 117   0   1   0]
 [  0   0   0   1   0 939   0  38   2  20]
 [ 91   6  72  80 104   0 633   0  14   0]
 [  0   0   0   0   0  34   0 933   0  33]
 [  3   1   2  15   6  24  33   5 910   1]
 [  0   1   0   0   0  17   1  33   0 948]]


In [80]:
incorrect_labels = np.where(y_test != pred_labels)[0]
print(f'The total number of incorrect labels are: {len(incorrect_labels)} \n')

print('Examples of Incorrect Classifications:')
for num in range(4):
    nums = incorrect_labels[num]
    print(f'Test: {y_test[nums]} Pred: {pred_labels[nums]}')

The total number of incorrect labels are: 1662 

Examples of Incorrect Classifications:
Test: 7 Pred: 5
Test: 9 Pred: 7
Test: 4 Pred: 2
Test: 0 Pred: 3


From these results we are able to recognize that there were 1662 False positives which could've impacted the outcome. The predicted data was not classified correctly to their corresponding test data